firmware/examples/Dac/source/main.cpp

```diff
@@ -0,0 +1,156 @@
+    222, 212, 202, 192, 183, 173, 164, 155,
+    146, 138, 130, 121, 113, 106, 98, 91,
+    84, 78, 71, 65, 59, 53, 48, 43,
+    38, 34, 29, 25, 22, 18, 15, 12,
+    10, 7, 5, 4, 2, 1, 1, 0,
+    0, 0, 1, 1, 2, 4, 5, 7,
+    10, 12, 15, 18, 22, 25, 29, 34,
+    38, 43, 48, 53, 59, 65, 71, 78,
+    84, 91, 98, 106, 113, 121, 130, 138,
+    146, 155, 164, 173, 183, 192, 202, 212,
+    222, 233, 243, 254, 264, 275, 286, 297,
+    309, 320, 332, 343, 355, 367, 379, 390,
+    402, 415, 427, 439, 451, 463, 475, 488
+};
+
+void SineDac(Dac * dac_demo, uint32_t delay)
+{
+    SJ2_ASSERT_FATAL(delay < 500,
+                     "Input frequency out of range! Frequency "
+                     "input needs to be between 0 and 500 Hz.");
+    for ( uint8_t i = 0; i < 255; i++ )
+    {
+        dac_demo->WriteDac(kSineLookup[i]);
+        Delay(delay);
+    }
+}
+void SawtoothDac(Dac * dac_demo, uint32_t delay)
+{
+    SJ2_ASSERT_FATAL(delay < 500,
+                     "Input delay out of range! Delay "
+                     "input needs to be between 0 and 500 Hz.");
+    int sawtooth = 0;
+    for ( uint8_t i = 0; i < 255; i++ )
+    {
+        uint16_t conversion = static_cast<uint16_t>(sawtooth);
+        dac_demo->WriteDac(conversion);
+        sawtooth = sawtooth + 4;
+        Delay(delay);
+    }
+}
+void TriangleDac(Dac * dac_demo, uint32_t delay)
+{
+    SJ2_ASSERT_FATAL(delay < 500,
+                     "Input delay out of range! Delay "
+                     "input needs to be between 0 and 500 Hz.");
+    int triangle = 0;
+    for ( uint8_t i = 0; i < 127; i++ )
+    {
+        uint16_t conversion = static_cast<uint16_t>(triangle);
+        dac_demo->WriteDac(conversion);
+        triangle = triangle + 8;
+        Delay(delay);
+    }
+    for ( uint8_t i = 127; i < 255 ; i++ )
+    {
+        uint16_t conversion = static_cast<uint16_t>(triangle);
+        dac_demo->WriteDac(conversion);
+        triangle = triangle - 8;
+        Delay(delay);
+    }
+}
+void SerratedDac(Dac * dac_demo, uint32_t delay)
+{
+    SJ2_ASSERT_FATAL(delay < 500,
+                     "Input delay out of range! Delay "
+                     "input needs to be between 0 and 500.");
+    int serrated = 0;
+    for ( uint8_t i = 0; i < 127; i++ )
+    {
+        uint16_t conversion = static_cast<uint16_t>(serrated);
+        dac_demo->WriteDac(conversion);
+        serrated = serrated + 8;
+        Delay(delay);
+    }
+    serrated = 0;
+    for ( uint8_t i = 127; i < 253 ; i++ )
+    {
+        uint16_t conversion = static_cast<uint16_t>(serrated);
+        dac_demo->WriteDac(conversion);
+        serrated = serrated + 4;
+        Delay(delay);
+    }
+}
+void (*dac_controllers[4])(Dac *, uint32_t) = {
+     SerratedDac,
+     TriangleDac,
+     SawtoothDac,
+     SineDac
+};
+void StartDemo(Dac dac_demo, uint32_t input_cycles)
+{
+    for (int j = 0; j < 60; j++)
+    {
+        dac_controllers[0](&dac_demo, input_cycles);
+    }
```

Comment:
Lol. So the reason why I said to use a look up table, is so you only have to create two for loops. You shouldn't be repreating them multiple times.

Reply:
```
for (int i = 0; i < sizeof(dac_controllers)/sizeof(*dac_controllers); i++)
{
    for (int j = 0; j < 60; j++)
    {
        dac_controllers[i](&dac_demo, input_cycles);
    }
}
```

Reply:
Can you also consider reducing the time in each waveform? Because it should be a quick demo someone can run through within a minute of having everything connected. 60 iterations is super long, especially when you are waiting to see the last waveform.